irohad/ametsuchi/impl/postgres_command_executor.cpp

```diff
@@ -244,18 +248,84 @@ namespace iroha {
-      soci::statement st = sql_.prepare
-          << "INSERT INTO account_has_roles(account_id, role_id) VALUES "
-             "(:account_id, :role_id)";
-      st.exchange(soci::use(account_id));
-      st.exchange(soci::use(role_name));
-      auto message_gen = [&] {
-        return (boost::format("failed to insert account role, account: '%s', "
-                              "role name: '%s'")
-                % account_id % role_name)
-            .str();
+      const auto bits = shared_model::interface::RolePermissionSet::size();
+      soci::statement st = sql_.prepare <<
+          R"(
+            WITH )"
+              + (is_genesis_ ? "" :
+                             R"(role_permissions AS (
+                SELECT permission FROM role_has_permissions
+                WHERE role_id = :role_id
+            ),
+            role_has_any_permission AS (
+                SELECT permission <> '0'::bit()"
+                         + std::to_string(bits) + R"() FROM role_permissions
+            ),
+            account_roles AS (
+                SELECT role_id FROM account_has_roles WHERE account_id = :creator_id
+            ),
+            account_has_role_permissions AS (
+                SELECT COALESCE(bit_or(rp.permission), '0'::bit()"
+                         + std::to_string(bits) + R"()) &
+                    (SELECT * FROM role_permissions) =
+                    (SELECT * FROM role_permissions)
+                FROM role_has_permissions AS rp
+                JOIN account_has_roles AS ar on ar.role_id = rp.role_id
+                WHERE ar.account_id = :creator_id
+            ),)")
+              +
+              R"( inserted AS (
+                INSERT INTO account_has_roles(account_id, role_id)
+                (
+                    SELECT :account_id, :role_id )"
+              + (is_genesis_
+                     ? ""
+                     :
+                     R"( WHERE (SELECT * FROM role_has_any_permission) AND
+                    EXISTS (SELECT * FROM account_roles) AND
+                    (SELECT * FROM account_has_role_permissions))")
+              +
+              R"() RETURNING (1)
+            )
+            SELECT CASE
+                WHEN EXISTS (SELECT * FROM inserted) THEN 0 )"
+              + (is_genesis_
+                     ? ""
+                     :
+                     R"( WHEN NOT (SELECT * FROM role_has_any_permission) THEN 1
+                WHEN NOT EXISTS (SELECT * FROM account_roles) THEN 2
+                WHEN NOT (SELECT * FROM account_has_role_permissions) THEN 3 )")
+              +
+              R"( ELSE 4
+            END AS result
+)";
+      st.exchange(soci::use(account_id, "account_id"));
+      st.exchange(soci::use(role_name, "role_id"));
+      st.exchange(soci::use(creator_account_id_, "creator_id"));
+      std::vector<std::function<std::string()>> message_gen = {
+          [&] {
+            return "is valid command validation failed: no permissions in role "
```

Comment:
do we need "is valid" prefix for each message?

Reply:
I think we don't